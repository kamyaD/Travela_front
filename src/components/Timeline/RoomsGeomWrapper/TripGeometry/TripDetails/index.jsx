import React, {Component} from 'react';
import {PropTypes} from 'prop-types';
import moment from 'moment';
import './TripDetails.scss';

export default class TripDetails extends Component {

  reduceDuration = (value, type, start, end) => {
    start = start.add(value, type);
    let diff = end.diff(start);
    return [start, diff];
  }

  _getDurationObject = (trip) => {
    let years, months, weeks, days;
    let start = moment(trip.checkInDate, 'YYYY-MM-DD');
    const end = moment(trip.checkOutDate, 'YYYY-MM-DD');
    if(!start.isValid() || !end.isValid())
      return 'n/a';
    // add 1 day for inclusivity
    let diff = end.add(1, 'days').diff(start);
    years = moment.duration(diff).years();
    // determine remainder  months
    [start, diff] = this.reduceDuration(years, 'years', start, end);
    months = moment.duration(diff).months();
    // determine remainder weeks
    [start, diff] = this.reduceDuration(months, 'months', start, end);
    weeks = moment.duration(diff).weeks();
    // determine remainder days
    [start, diff] = this.reduceDuration(weeks, 'weeks', start, end);
    days = moment.duration(diff).days();
    return {years, months, weeks, days};
  }

  humanizeDuration = (trip) => {
    const values = this._getDurationObject(trip);
    if (values === 'n/a')
      return values;
    const durationStrings = Object.keys(values).map(key => {
      if(!values[key])
        return '';
      let capitalizedKey = key.charAt(0)
        .toUpperCase()
        .concat(key.substring(1));
      if(values[key] === 1)
        capitalizedKey = capitalizedKey.slice(0, -1);

      return `${values[key]} ${capitalizedKey}`;
    });
    const humanFriendlyString = durationStrings.join(' ');
    return humanFriendlyString;
  }

  handleClickBookingDetailsBody = (e) => {
    e.stopPropagation();
  }

  handleCloseDetailsButton = () => {
    const {toggleBookingDetails} = this.props;
    toggleBookingDetails('close');
  }

  renderBookingDetailsHeader = (trip) => {
    return (
      <div className="trip-booking-details__header">
        <span>{trip.request.name}</span>
        <div
          tabIndex="-1"
          role="presentation"
          onKeyUp={()=>{}}
          className="close-button"
          onClick={this.handleCloseDetailsButton}
        >
          <div>&times;</div>
        </div>
      </div>
    );
  }

  renderBookingDetailsBody = (trip) => {
    const checkInDate = trip.checkInDate
      ? moment(trip.checkInDate, 'YYYY-MM-DD').format('DD MMM YYYY')
      : 'n/a';
    const checkOutDate = trip.checkOutDate
      ? moment(trip.checkOutDate, 'YYYY-MM-DD').format('DD MMM YYYY')
      : 'n/a';
    const { handleChangeRoomModal } = this.props;

    return (
      <div className="trip-booking-details__body">
        <div>
          <span>Status</span>
          <span>Booked</span>
        </div>
        <div>
          <span>Check-in</span>
          <span>{checkInDate}</span>
        </div>
        <div>
          <span>Check-out</span>
          <span>{checkOutDate}</span>
        </div>
        <div>
          <span>Duration</span>
          <span>{this.humanizeDuration(trip)}</span>
        </div>
        <div>
          <span>Origin</span>
          <span>{trip.origin}</span>
        </div>
        <div>
          <span>
            <button
              type="button"
              className="trip-booking-details__button"
              onClick={() => handleChangeRoomModal(trip)}
            >
          Change Room
            </button>
          </span>
        </div>
      </div>
    );
  }

  determineTranslateClass = () => {
    const { translateDetailsLeft } = this.props;
    return translateDetailsLeft ? 'translateX' : '';
  }

  render() {
    const { bookingDetailsPos, detailsVariantClass, trip } = this.props;
    const translateClass = this.determineTranslateClass();
    return (
      <div
        className={`trip-booking-details details-${detailsVariantClass} ${translateClass}`}
        onClick={this.handleClickBookingDetailsBody}
        role="presentation"
        style={{
          position: 'absolute',
          left: `${bookingDetailsPos}px`,
        }}
      >
        {this.renderBookingDetailsHeader(trip)}
        {this.renderBookingDetailsBody(trip)}
      </div>
    );
  }
}

TripDetails.propTypes = {
  trip: PropTypes.object.isRequired,
  bookingDetailsPos: PropTypes.string,
  detailsVariantClass: PropTypes.string,
  toggleBookingDetails: PropTypes.func.isRequired,
  translateDetailsLeft: PropTypes.bool,
  handleChangeRoomModal: PropTypes.func.isRequired
};

TripDetails.defaultProps = {
  bookingDetailsPos: '100px',
  detailsVariantClass: 'hidden',
  translateDetailsLeft: false
};
