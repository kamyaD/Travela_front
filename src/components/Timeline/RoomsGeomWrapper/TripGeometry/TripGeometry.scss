.room-geometry-wrapper {
  div.timeline-trip-geometry {
    position: absolute;
    display: inline-block;
    cursor: pointer;
    outline: none;

    .geom-trip {
      height: 21px;
      border-radius: 12px;
      outline: none;

      &--outer {
        display: flex;
        align-items: center;
        position: relative;
        transition: width .5s cubic-bezier(.17,.67,.35,.05);
        overflow: hidden;
        outline: none;
      }
      &--inner + span {
        position: absolute;
        left: 20px;
        color: #FFFFFF;
        font-size: 12px;
        font-weight: 500;
        overflow-x: hidden;
        width: 150%;
      }
    }
  }
}
