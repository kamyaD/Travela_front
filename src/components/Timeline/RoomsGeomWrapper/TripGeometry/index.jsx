import React, {Component, Fragment} from 'react';
import {PropTypes} from 'prop-types';
import moment from 'moment';
import TripDetails from './TripDetails';
import TripGeomHelper from './TripGeomHelper';
import {COLOR_SPEC_RANGES, SHADE_OFFSETS} from '../../settings';
import './TripGeometry.scss';

class TripGeometry extends Component {
  constructor(props) {
    super(props);
    this.state = {
      bookingDetailsVisible: false,
      colorSpecs: this.shuffleColorSpecs(),
      bookingDetailsPos: 0,
      translateDetailsLeft: false
    };
  }

  checkShouldTranslateDetails = (mouseX) => {
    const screenWidth = window.innerWidth;
    if(screenWidth/2 < mouseX)
      return true;
    return false;
  }

  setBookingDetailsAbsolutePos = (e) => {
    const tripGeomLeft = e.target.getBoundingClientRect().left;
    const mouseX = e.clientX;
    const bookingDetailsPos = mouseX - tripGeomLeft;
    const translateDetailsLeft = this.checkShouldTranslateDetails(mouseX);
    this.setState(prevState => ({
      ...prevState,
      bookingDetailsPos,
      translateDetailsLeft
    }));
  }

  toggleBookingDetails = (type) => {
    this.setState(prevState => {
      switch (type) {
      case 'open':
        if(prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: true
        };
      default:
        if(!prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: false
        };
      }
    });
  }

  getInnerColorShade = (colorSpecs) => {
    let {hue, saturation, lightness} = colorSpecs;
    saturation+=SHADE_OFFSETS.saturation;
    lightness+=SHADE_OFFSETS.lightness;
    return {hue, saturation, lightness};
  }

  pickValueFromRange = (colorSpec) => {
    // add a random fraction of the magnitude to min value
    return Math.round(colorSpec.min + (
      Math.random() * (colorSpec.max - colorSpec.min)
    ));
  }

  shuffleColorSpecs = () => {
    const hue = this.pickValueFromRange(COLOR_SPEC_RANGES.hue);
    const saturation = this.pickValueFromRange(COLOR_SPEC_RANGES.saturation);
    const lightness = this.pickValueFromRange(COLOR_SPEC_RANGES.lightness);
    return {hue, saturation, lightness};
  }

  getBookingDetailsVariant = () => {
    const {bookingDetailsVisible} = this.state;
    return bookingDetailsVisible ? 'visible' : 'hidden';
  }

  renderTripBarInnerElements(tripStats, colorSpecs) {
    const {trip} = this.props;
    const {hue, saturation, lightness} = this.getInnerColorShade(colorSpecs);
    const travellerName = trip.request.name;
    const travellerGenderLabel = trip.request.gender.charAt(0);
    return (
      <Fragment>
        <div
          className="geom-trip geom-trip--inner"
          style={{
            width: tripStats.stayPercentage,
            background: `hsl(${hue},${saturation}%,${lightness}%)`
          }}
        />
        <span>{`${travellerName}  (${travellerGenderLabel})`}</span>
      </Fragment>
    );
  }

  renderTripBar = (tripStats) => {
    const {colorSpecs} = this.state;
    const {tripDayWidth} = this.props;
    const {hue, saturation, lightness} = colorSpecs;
    return (
      <div
        className="geom-trip geom-trip--outer"
        style={{
          minHeight: '20px',
          background: `hsl(${hue},${saturation}%,${lightness}%)`,
          width: `${tripDayWidth * tripStats.length}px`
        }}
      >
        {this.renderTripBarInnerElements(tripStats, colorSpecs)}
      </div>
    );
  }

  render() {
    const {bookingDetailsPos, translateDetailsLeft} = this.state;
    const {trip, handleChangeRoomModal} = this.props;
    const detailsVariant = this.getBookingDetailsVariant();
    const helper = new TripGeomHelper(this.props);
    const tripStats = helper.getTripStats(trip);
    return (
      <div
        className="timeline-trip-geometry"
        tabIndex="-1"
        role="presentation"
        onKeyUp={()=>{}}
        onClick={this.setBookingDetailsAbsolutePos}
        onFocus={() => this.toggleBookingDetails('open')}
        onBlur={() => this.toggleBookingDetails('close')}
        style={{
          left: `${tripStats.tripTimelineOffsetLeft}px`
        }}
      >
        {this.renderTripBar(tripStats)}
        <TripDetails
          trip={trip}
          bookingDetailsPos={bookingDetailsPos}
          translateDetailsLeft={translateDetailsLeft}
          detailsVariantClass={detailsVariant}
          toggleBookingDetails={this.toggleBookingDetails}
          handleChangeRoomModal={handleChangeRoomModal}
        />
      </div>
    );
  }
}

TripGeometry.propTypes = {
  trip: PropTypes.object.isRequired,
  tripDayWidth: PropTypes.number.isRequired,
  handleChangeRoomModal: PropTypes.func.isRequired
};

export default TripGeometry;
