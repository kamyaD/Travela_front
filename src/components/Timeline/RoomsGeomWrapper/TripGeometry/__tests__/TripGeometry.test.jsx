import React from 'react';
import TripGeometry from '..';

const props = {
  trip: {
    departureDate: '2018-01-01',
    returnDate: '2018-01-07',
    request: {
      name: 'John Doe',
      gender: 'Male'
    }
  },
  tripDayWidth: 31
};

const bookingDetailsPosEvent = {
  target: {
    getBoundingClientRect: jest.fn(() => ({
      left: 200,
    }))
  },
  clientX: 300
};

describe('<TripGeometry />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<TripGeometry {...props} />);
  });


  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('timeline trip geometry', () => {
    let timelineBarWrapper, blurFocusSpy;
    beforeEach(() => {
      timelineBarWrapper = wrapper.find('TimelineBarWrapper');
      blurFocusSpy = jest.spyOn(wrapper.instance(), 'toggleBookingDetails');
    });

    it('renders the timeline trip geometry', () => {
      expect(timelineBarWrapper).toHaveLength(1);
    });

    describe('booking details toggling', () => {
      let onFocusHandler, onBlurHandler, bookingDetailsPosHandler;
      beforeEach(() => {
        onFocusHandler = timelineBarWrapper
          .dive()
          .find('.timeline-trip-geometry')
          .prop('onFocus');
        onBlurHandler = timelineBarWrapper
          .dive()
          .find('.timeline-trip-geometry')
          .prop('onBlur');
        bookingDetailsPosHandler = timelineBarWrapper
          .prop('setBookingDetailsAbsolutePos');
      });

      it('lauches with booking details hidden', () => {
        expect(wrapper.state().bookingDetailsVisible).toBe(false);
      });

      describe('toggle booking details handler works properly', () => {
        it('leaves the booking details open when clicked severally', () => {
          expect.assertions(2);
          onFocusHandler(); //click
          expect(wrapper.state().bookingDetailsVisible).toBe(true);
          onFocusHandler(); // click once more
          expect(wrapper.state().bookingDetailsVisible).toBe(true);
        });

        it('closes visible booking details on blur', () => {
          expect.assertions(2);
          onFocusHandler(); // open booking details
          expect(wrapper.state().bookingDetailsVisible).toBe(true);
          onBlurHandler(); // close booking details
          expect(wrapper.state().bookingDetailsVisible).toBe(false);
        });

        it('passes correct absolute position to booking details', () => {
          bookingDetailsPosHandler(bookingDetailsPosEvent);
          // mouseLeft - tripLeft => see event object
          const expectedBookingDetailsPos = 100;
          expect(wrapper.state().bookingDetailsPos)
            .toEqual(expectedBookingDetailsPos);
          const tripDetails = wrapper.find('TripDetails');
          expect(tripDetails.prop('bookingDetailsPos'))
            .toBe(expectedBookingDetailsPos);
        });
      });
    });

    it('renders the TripDetails elements', () => {
      const tripDetails = wrapper.find('TripDetails');
      expect(tripDetails).toHaveLength(1);
    });
  });
});
