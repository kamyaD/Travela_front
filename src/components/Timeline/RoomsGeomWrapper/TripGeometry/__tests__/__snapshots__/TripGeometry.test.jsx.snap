// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripGeometry /> renders correctly 1`] = `
<TimelineBarWrapper
  setBookingDetailsAbsolutePos={[Function]}
  toggleBookingDetails={[Function]}
  tripStats={
    Object {
      "length": 7,
      "stayPercentage": "100%",
      "tripTimelineOffsetLeft": -11811,
    }
  }
>
  <TimelineBar
    customStyle={
      Object {
        "background": "hsl(227,71%,58%)",
      }
    }
    tripDayWidth={31}
    tripStats={
      Object {
        "length": 7,
        "stayPercentage": "100%",
        "tripTimelineOffsetLeft": -11811,
      }
    }
  >
    <div
      className="geom-trip geom-trip--inner"
      style={
        Object {
          "background": "hsl(227,81%,38%)",
          "width": "100%",
        }
      }
    />
    <span>
      John Doe  (M)
    </span>
  </TimelineBar>
  <TripDetails
    bookingDetailsPos={0}
    detailsVariantClass="hidden"
    handleChangeRoomModal={[Function]}
    toggleBookingDetails={[Function]}
    translateDetailsLeft={false}
    trip={
      Object {
        "departureDate": "2018-01-01",
        "request": Object {
          "gender": "Male",
          "name": "John Doe",
        },
        "returnDate": "2018-01-07",
      }
    }
  />
</TimelineBarWrapper>
`;
